include/deal.II/matrix_free/operators.h

```diff
@@ -1091,23 +1114,75 @@ namespace MatrixFreeOperators
           const VectorizedArrayType *               in_array,
           VectorizedArrayType *                     out_array) const
   {
-    const unsigned int given_degree =
-      fe_eval.get_shape_info().data[0].fe_degree;
     if (fe_degree > -1)
-      internal::CellwiseInverseMassMatrixImplFlexible<dim,
-                                                      VectorizedArrayType>::
-        template run<fe_degree>(
-          n_actual_components,
-          fe_eval.get_shape_info().data.front().inverse_shape_values_eo,
-          inverse_coefficients,
-          in_array,
-          out_array);
+      internal::CellwiseInverseMassMatrixImplFlexible<
+        dim,
+        VectorizedArrayType>::template run<fe_degree>(n_actual_components,
+                                                      fe_eval,
+                                                      inverse_coefficients,
+                                                      false,
+                                                      in_array,
+                                                      out_array);
+    else
+      internal::CellwiseInverseMassFactory<dim, VectorizedArrayType>::apply(
+        n_actual_components,
+        fe_eval,
+        inverse_coefficients,
+        false,
+        in_array,
+        out_array);
+  }
+
+  template <int dim,
+            int fe_degree,
+            int n_components,
+            typename Number,
+            typename VectorizedArrayType>
+  inline void
+  CellwiseInverseMassMatrix<dim,
+                            fe_degree,
+                            n_components,
+                            Number,
+                            VectorizedArrayType>::
+    apply(const AlignedVector<Tensor<2, n_components, VectorizedArrayType>>
+            &                        inverse_dyadic_coefficients,
+          const unsigned int         n_actual_components,
+          const VectorizedArrayType *in_array,
+          VectorizedArrayType *      out_array) const
+  {
+    const unsigned int dofs_per_component = inverse_dyadic_coefficients.size();
+    const unsigned int n_tensor_components =
+      n_actual_components * n_actual_components;
+
+    AlignedVector<VectorizedArrayType> inverse_coefficients(
+      dofs_per_component * n_tensor_components);
+
+    // Flatten the inverse dyadic coefficients into `inverse_coefficients`
+    {
+      auto begin = inverse_coefficients.begin();
+      for (unsigned int q = 0; q < dofs_per_component; ++q)
+        {
+          const auto end = std::next(begin, n_tensor_components);
+          inverse_dyadic_coefficients[q].unroll(begin, end);
+          begin = end;
+        }
+    }
```

Comment:
Do we need this code? Apart from the fact that I would like to avoid the memory allocation of `AlignedVector`, this also seems like unnecessary work. I would have simply passed in a `Number*` or `VectorizedArray*` to the internal `run` functions. This forces you to also pass in the size, since you can't call `inverse_coefficients.size()` any more. Of course, working with plain pointers seems less nice from a programming point of view, but these are performance critical functions, so we should not copy data around (and spill caches etc). We could of course opt to use `ArrayView<const Number>` as the argument to `run()`, which allows to query the size (in fact, it is precisely a pointer to the start and the size): https://github.com/dealii/dealii/blob/947199be8142d3fc7bde15d8f94f168573ccf5d2/include/deal.II/base/array_view.h#L368-L377

Reply:
@kronbichler See discussion above (https://github.com/dealii/dealii/pull/14850#pullrequestreview-1327861878 and https://github.com/dealii/dealii/pull/14850#issuecomment-1457901809).

Reply:
How do we want to proceed here?